### PR TITLE
test: some improvements

### PR DESCRIPTION
**Pull request by TimoGlastra**
We're having a lot of issues with tests randomly failing. I think the issue is in the in memory wallet being faster (espeically in closing etc..) and therefore the closing / shutdown after tests is now going wrong when we use message pickup.

This PR adds some small improvements, but I think the proper way to fix this is by keeping track of which pickup messages we have sent, and waiting for a response to the sent messages after stopMessagePickup has been called. That way we know for sure no more messages will be exchanged before we call `shutdown` on the agents.

Partially fixes https://github.com/openwallet-foundation/credo-ts/issues/1750 It moves more to in-memory. splits e2e from non-e2e, splits non-e2e into shards (2 at the moment) to speed up tests. And I made a first pass at using pre-generated anoncreds definitions in tests. The tests i did this for were significantly faster

### CODE REVIEW

**Review by genaris — APPROVED** [review]
Great additions here! This pre-creation of AnonCreds objects will significantly reduce the risks of tests failing due to timeout errors (apart from making them faster, as you've already benchmarked).

I've left some minor comments, nothing really significant besides a doubt I have regarding the closing of `WsOutboundTransport` that may introduce issues if works as I remember it did in nodejs environment (probably it has changed since the time I checked).

File: packages/anoncreds/tests/preCreatedAnonCredsDefinition.ts [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+      ],
+      [
+        'profile_picture',
+        '626128909514031861048611274299543488396178512804450237734576430023887274743970983888968738560062281447232501644235732190657881416295947662076218443595968416492300135536820604684825587002869836933465348240667269380721505682854695936658536043347142984963045658919215661548340875420653437468343448949143330848700282622905251500903165366616989093528387548421898172630311164792908296732119373995492469657592915994374916970385323733879692921481867888839992919173543562249790190392930924209699652327647797204082384416540630932657082008347095070625889215957989868906105233598674999715503418316871508439518805591147670972038744125173028393048388065281173594232476773594862090167445272672533232951411418',
+      ],
+    ],
+  },
+} satisfies PreCreatedAnonCredsDefinition
+
+export async function usePreCreatedAnonCredsDefinition(
```
[the remark this answers is not in the document]
Comment (TimoGlastra) [reply packages/anoncreds/tests/preCreatedAnonCredsDefinition.ts:98]: True!! 😅 

Changed it to `storePreCreatedAnonCredsDefinition`

File: packages/core/src/modules/oob/__tests__/implicit.test.ts [hunk 2]
```diff
@@ -219,50 +211,67 @@ describe('out of band implicit', () => {
...
     expect(connectionsFromFaberPublicDid).toEqual(
       expect.arrayContaining([aliceFaberConnection, aliceFaberNewConnection])
     )
   })
 })
 
-async function createPublicDid(agent: Agent, unqualifiedSubmitterDid: string, endpoint: string) {
-  const createResult = await agent.dids.create<IndyVdrDidCreateOptions>({
-    method: 'indy',
+async function createPeerDid(agent: Agent, endpoint: string) {
```
[the remark this answers is not in the document]
Comment (TimoGlastra) [reply packages/core/src/modules/oob/__tests__/implicit.test.ts:225]: Agreed! Changed it to use `did:inmemory`. Cleaner also now :)

File: jest.config.base.ts [hunk 3]
```diff
@@ -3,7 +3,9 @@ import type { Config } from '@jest/types'
 const config: Config.InitialOptions = {
   preset: 'ts-jest',
   testEnvironment: 'node',
-  testMatch: ['**/?(*.)+(spec|test).[tj]s?(x)'],
+  // NOTE: overwritten in e2e test. Make sure to
```
[the remark this answers is not in the document]
Comment (TimoGlastra) [reply jest.config.base.ts:6]: I think both can be used in this case, 😄 . But will change! 

https://chat.openai.com/share/59212a33-041e-4542-956b-bdb7fd65de30

File: tests/e2e-test.ts [hunk 4]
```diff
@@ -109,5 +129,33 @@ export async function e2eTest({
 
   // We want to stop the mediator polling before the agent is shutdown.
   await recipientAgent.mediationRecipient.stopMessagePickup()
-  await sleep(2000)
+
+  const pickupRequestMessages = [V2DeliveryRequestMessage.type.messageTypeUri, V1BatchPickupMessage.type.messageTypeUri]
```
[the remark this answers is not in the document]
Comment (TimoGlastra) [reply tests/e2e-test.ts:133]: I think so yes, but we would still need to hook into the sent event to catch the thread id that was used for the last sent message.

I think we should probably also hook the initateMessagePickup into this and the subscription from initiateMessagePickup so `stopMessagePickup` will return when the message pickup is fully complete. If a pickup message has just been sent, it should wait for the delivery message before resolving. I think if we can do that in a follow up, It'll allow us to clean up quite some test logic related to shutting down at the correct time

File: packages/core/src/transport/WsOutboundTransport.ts [hunk 5]
```diff
@@ -27,14 +29,26 @@ export class WsOutboundTransport implements OutboundTransport {
...
     this.logger.debug('Stopping WS outbound transport')
+    this.isActive = false
+
+    const stillOpenSocketClosingPromises: Array<Promise<void>> = []
+
     this.transportTable.forEach((socket) => {
       socket.removeEventListener('message', this.handleMessageEvent)
-      socket.close()
+      if (socket.readyState !== WebSocket.CLOSED) {
+        stillOpenSocketClosingPromises.push(new Promise((resolve) => socket.once('close', resolve)))
```
[the remark this answers is not in the document]
Comment (TimoGlastra) [reply packages/core/src/transport/WsOutboundTransport.ts:45]: I've just checked and it is called when close is iniated by us.

**Comment by TimoGlastra** [comment]
Merging for now. Thanks for the feedback! We can improve the pickup finished handling in a follow up PR (if we add the right handles in the `stopMessagePickup`)

**Pull request merged**